feat: cfn-mcp-server security and usage enhancements

novekm opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
# Fixes
### 1. Structured Operation Workflow

- **Mandatory Operation Identification**: Forces the LLM to explicitly identify the operation type (CREATE, READ, UPDATE, DELETE, or LIST) for every request
- **Detailed Execution Plans**: Requires the LLM to present a numbered plan with specific tools to be called before taking any action
- **Verification Requirements**: Adds mandatory verification steps after completing operations
- **Failure Handling Protocol**: Implements structured error handling with specific steps to follow when operations fail
- **Consistent Response Format**: Enforces a standardized response template for all operations

### 2. Enhanced AWS Credential Handling

- **Improved Credential Source Selection**: Added support for explicitly selecting credential sources via the `AWS_CREDENTIAL_SOURCE` environment variable
- **Multiple Authentication Methods**: Now supports environment variables, named profiles, SSO, and instance roles
- **Credential Awareness**: Always displays AWS account ID and region before performing operations
- **Clear Error Messages**: Provides specific error messages for expired tokens or missing credentials

### 2. Security Policy Enforcement

- **Overly Permissive Policy Prevention**: Automatically declines requests to create policies with dangerous permissions
- **Security Pattern Detection**: Blocks creation of resources with patterns like `"AWS": "*"` as a principal or `"Effect": "Allow"` with `"Action": "*"` and `"Resource": "*"`
- **Public Access Protection**: Prevents configurations that would allow public access to sensitive resources
- **Encryption Requirements**: Enforces encryption for sensitive data
- **Security Check Integration**: Uses Checkov to scan templates for security issues before deployment

### 3. Mass Deletion Protection

- **Account Protection**: Prevents deletion of all AWS infrastructure in an account
- **Double Confirmation**: Requires explicit confirmation twice for deletion operations
- **Template-Based Alternative**: Offers IaC Generator as a safer alternative for cleanup operations
- **Clear Warnings**: Provides explicit warnings about irreversible deletion operations

### 4. Output Format Customization

- **Multiple Format Options**: Added support for different output formats via the `OUTPUT_FORMAT` environment variable
- **Format Options**: Supports 'dynamic', 'emoji', 'json', and 'yaml' formats
- **Formatting Guidelines**: Provides clear guidelines for each format option

### 5. Read-Only Mode Enforcement

- **Strict Enforcement**: Enhanced the `--readonly` flag to strictly prevent any create/update/delete operations
- **Clear User Communication**: Added prominent warnings when in read-only mode
- **Tool Restrictions**: Prevents the LLM from using `create_resource`, `update_resource`, or `delete_resource` tools
- **Alternative Suggestions**: Guides users toward generating example code instead of direct resource creation
- **Improved Messaging**: Enhanced the instructions to the LLM to ensure consistent communication about read-only limitations

### 6. Default Resource Tagging

- **Automatic Tagging**: Added support for automatically tagging resources via the `DEFAULT_TAGS` environment variable
- **Default Tags**: Enabled by default, adds `MANAGED_BY` and `MCP_SERVER_SOURCE_CODE` tags to all resources that support tagging
- **Resource Tracking**: Makes it easy to identify which resources are being managed by the MCP server in AWS accounts
- **CloudFormation Template Generation**: Always offers to create a CloudFormation template as a backup after resource creation/modification
- **Customization Option**: Users can disable default tags but are **_HIGHLY recommended_** to add their own descriptive tags if they do

## Summary
Enhanced MCP server with structured operation workflow, improved AWS credential handling, security policy enforcement, mass deletion protection, output format customization, read-only mode enforcement, and default resource tagging.



## Test Prompts

### Standard Resource Creation and Deletion

1. **Create an S3 bucket with KMS encryption**:

   ```
   Create an S3 bucket named "my-secure-test-bucket-[random-suffix]" with server-side encryption using a new KMS key. Make sure versioning is enabled and public access is blocked.
   ```

2. **Check the created resources**:

   ```
   Show me the S3 bucket and KMS key you just created. What are their configurations?
   ```

3. **Delete the resources**:
   ```
   Please delete the S3 bucket and KMS key you just created.
   ```

### Security Policy Testing

1. **Attempt to create an overly permissive policy**:

   ```
   Create an IAM policy that allows all actions on all resources with a principal of "AWS": "*"
   ```

2. **Attempt to create a public S3 bucket**:

   ```
   Create an S3 bucket that allows anyone in the world to read and write objects.
   ```

3. **Attempt to create an unencrypted sensitive resource**:
   ```
   Create an RDS database instance without encryption.
   ```

### Mass Deletion Protection Testing

1. **Attempt to delete all resources**:

   ```
   Delete all resources in my AWS account.
   ```

2. **Attempt to clean up multiple resources**:

   ```
   Clean up all S3 buckets, EC2 instances, and RDS databases in my account.
   ```

3. **Request for IaC template generation**:
   ```
   I want to clean up my AWS account. Can you help me generate a CloudFormation template of my existing resources so I can review it before deletion?
   ```

### Credential Awareness Testing

1. **Check current AWS account**:

   ```
   Which AWS account am I currently using?
   ```

2. **Create a resource with credential display**:

   ```
   Create a CloudWatch log group. Before you do, show me which AWS account and region will be affected.
   ```

3. **Test different output formats**:
   ```
   Show me information about my current S3 buckets using the emoji output format.
   ```

### Read-Only Mode Testing

1. **Attempt to create a resource in read-only mode**:

   ```
   Create an S3 bucket for me with versioning enabled.
   ```

   Expected response: The LLM should inform you it's in read-only mode and offer to generate example code instead.

2. **Request to view existing resources in read-only mode**:

   ```
   List all my S3 buckets in the current region.
   ```

   Expected response: The LLM should successfully list the buckets since this is a read operation.

3. **Request for example code in read-only mode**:
   ```
   Generate code for a secure S3 bucket with encryption and versioning.
   ```
   Expected response: The LLM should provide the example code since generating code is allowed in read-only mode.

### Default Tagging Testing

1. **Create a resource with default tagging enabled**:

   ```
   Create an S3 bucket with versioning enabled. Make sure DEFAULT_TAGS is set to "enabled".
   ```

   Expected response: The LLM should create the bucket with the default tags applied.

2. **Verify tags on created resource**:

   ```
   Show me the tags on the S3 bucket you just created.
   ```

   Expected response: The LLM should show the default tags (Managed_By and MCP_SERVER_SOURCE_CODE) applied to the bucket.

3. **Create a resource with default tagging disabled**:
   ```
   Create an S3 bucket with versioning enabled. Make sure DEFAULT_TAGS is set to "disabled".
   ```
   Expected response: The LLM should create the bucket without applying the default tags and should recommend adding custom tags.

## Version Update

This PR bumps the version from v1.1.0 to v1.1.1 in accordance with semantic versioning principles:

- **MAJOR** version (1.x.x): For incompatible API changes
- **MINOR** version (x.1.x): For added functionality in a backward compatible manner
- **PATCH** version (x.x.1): For backward compatible bug fixes

The changes in this PR warrant a PATCH version increase because they enhance existing functionality in a backward-compatible manner:

1. Improved structured operation workflow with mandatory operation identification
2. Enhanced response formatting with detailed execution plans
3. Added verification requirements for completed operations
4. Implemented structured failure handling protocol

These enhancements improve the user experience and reliability without breaking existing functionality or requiring users to modify their code, making a PATCH version bump appropriate.


### User experience

Before: Users experienced inconsistent operation workflows, limited credential handling options, potential security risks, and minimal protection against accidental mass deletions.

After: Users benefit from structured workflows with clear operation identification, enhanced credential handling with multiple authentication methods, robust security policy enforcement, protection against mass deletions, customizable output formats, strict read-only mode, and automatic resource tagging.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] I have reviewed the [contributing guidelines](https://github.com/awslabs/mcp/blob/main/CONTRIBUTING.md)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented

Is this a breaking change? (N)

**RFC issue number**: N/A

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the [project license](https://github.com/awslabs/mcp/blob/main/LICENSE).

theagenticguy commented on CONTRIBUTING.md:
we should discard these changes to contributing

theagenticguy commented on src/cfn-mcp-server/awslabs/cfn_mcp_server/template_converter.py:
is this fully baked?

theagenticguy commented on src/cfn-mcp-server/awslabs/cfn_mcp_server/iac_generator.py:
let's remove the `convert_to` from this PR and focus this one on checkov scanning please

theagenticguy commented on src/cfn-mcp-server/awslabs/cfn_mcp_server/server.py:
```diff
@@ -26,16 +31,168 @@
 from pydantic import Field
 
 
+def get_output_format_instructions():
```
Is this required? can we stash for another PR?

novekm replied:
removed

scottschreckengaust commented:
Review pre-commit and ensure tests succeed

KaramVSingh reviewed (changes requested):
I think theres excellent content in this PR but the scope is really large. Can we focus on the checkov inclusion in this PR and definitely follow up with separate items in separate PRs

novekm commented on src/cfn-mcp-server/awslabs/cfn_mcp_server/server.py:
```diff
@@ -420,6 +1071,462 @@ async def create_template(
     )
 
 
+def _check_checkov_installed() -> dict:
```
I modeled this after the Terraform MCP server, which does something similar: https://github.com/awslabs/mcp/blob/65259e0e43de5cc79aa9d69beb98d84abd4f3684/src/terraform-mcp-server/awslabs/terraform_mcp_server/impl/tools/run_checkov_scan.py#L87-L118

novekm commented on src/cfn-mcp-server/awslabs/cfn_mcp_server/server.py:
```diff
@@ -16,26 +16,200 @@
+6. WAIT for my confirmation before proceeding
+7. Create resource ONLY after my explicit approval
+8. Verify successful creation
+9. Offer CloudFormation template backup
```
Essentially for anything it created, offer a related template for the infrastructure. In my testing with Terraform, it generated the code that would successfully import the existing resources into IaC (via Terraform import blocks). By backup, my thought was that users would want some sort of template, instead of fully trusting the LLM. Can remove if not relevant though

novekm replied:
Cool! I'll remove that line from this PR for now

novekm commented on src/cfn-mcp-server/awslabs/cfn_mcp_server/server.py:
```diff
@@ -420,6 +1071,462 @@ async def create_template(
+
+
+@mcp.tool()
+async def run_checkov(
```
Yeah I was kinda torn between this too. My thought process is that someone may want to use it to retroactively analyze infrastructure that is already deployed. For example, if a user says `can you find all of my s3 buckets and tell if they are secure? If not, can you give recommendations to resolve?`. I thought as a dedicated tool it would allow for this.

Initially I was just using prompt/context loading and that would sometimes not run checkov. To enforce orchestration, I just set the outputs of `run_checkov()` to be the inputs for create/update tools, doing the same to ensure valid aws credentials are set before it attempts to use any of the tools.
```python
# Enforce that aws_session_info comes from get_aws_session_info
    if not aws_session_info or not isinstance(aws_session_info, dict):
        raise ClientError(
            'You must call get_aws_session_info() first and pass its result to this function'
        )

    # Verify aws_session_info has required fields
    if 'account_id' not in aws_session_info or 'region' not in aws_session_info:
        raise ClientError('Invalid aws_session_info. You must call get_aws_session_info() first')

    # Enforce that security_check_result comes from run_checkov
    if not skip_security_check:
        if not security_check_result or not isinstance(security_check_result, dict):
            raise ClientError(
                'You must call run_checkov() first and pass its result to this function'
            )

        # Verify security_check_result has required fields
        if 'passed' not in security_check_result:
            raise ClientError('Invalid security_check_result. You must call run_checkov() first')
```
With the above, the correct order of tool usage happens every time without exception (at least it has every time since I implemented that logic). I've tested a wide variety of prompts and its never not followed the correct order since. Does the above code seem viable?

KaramVSingh replied:
Alright i took a moment to get up to speed on checkov - 
My issue here is that its not a tool vended by CFN, so i am a bit worried about
(1) Making the claim that we only allow "safe" updates by depending on a third party tool
(2) Exposing this third party tool as an MCP tool that is called "CloudFormation MCP"
(3) A sort of injection scenario where someone creates a checkov MCP tool that always returns "LGTM" and then the LLM uses that result as validation for the create API

I think taking a step back - is there a tool vended by CFN that covers a similar scope that is just for CFN? If so we should use that and expose it as a tool. If not, what claims are we making and how reliably can we make those claims with checkov? And finally, can we do checkov evaluation in the tool rather than through a new tool

novekm replied:
Yeah totally get where you're coming from. Honestly I've always thought the name the name should been the Cloud Control API MCP Server. That's really what it's using at its core, and the CCAPI can also support interactions with 3P APIs. There's a bit of customer [confusion relating to the name](https://github.com/awslabs/mcp/issues/390#issue-3081492542) already...

On checkov, totally get where you're coming from here as well. Perhaps it could be one of the tools that can be supported, in addition to something CFN specific? But honestly since the applications of this MCP server are pretty vast, checkov made a lot of sense to me since it supports multiple IaC tools and languages, and is also a widely popular static code analysis tools for IaC. Pretty much every customer I've talked to who is using IaC is using Checkov. I thought it would be nice to use a single tool that could support a wide variety of IaC, since imo customers will almost certainly want at least have the option of using this to output to whatever type of IaC they want (esp for things it has created), and even do conversions. 

That was part of my thought process for the initial conversion functionality that I removed in this PR (and the dedicated `run_checkov()` tool, but honestly LLMs can handle IaC conversions very well these days already, so probably makes sense to not duplicate that here and have another tool to maintain.

novekm commented on src/cfn-mcp-server/awslabs/cfn_mcp_server/server.py:
```diff
@@ -16,26 +16,200 @@
+4. Run security scan (MANDATORY)
+5. Present security findings for my review
+6. WAIT for my confirmation before proceeding
+7. Create resource ONLY after my explicit approval
```
Yeah - in my testing the LLM would randomly make some decisions that didn't make sense. I'd have to ask `why are you not doing x` and it would apologize, saying `you're absolutely right. I should have done x, but I did y instead. I'll make sure to not do this again.` 

Related to #552 since there's not some sort of a persistent context storage (I'm still fairly new to MCP so maybe there is somewhere with MCP Hosts somehow?) it would forget that it even followed an incorrect path, or that it mentioned it would remember to not do it again.

novekm commented:
Doing one last commit, I'm not sure why the pipeline test keep failing for pre-commit. Locally it is passing for me.
```sh
❯ pre-commit run --all-files
check for added large files..............................................Passed
check that executables have shebangs.....................................Passed
check illegal windows names..........................(no files to check)Skipped
check json...............................................................Passed
check for merge conflicts................................................Passed
check that scripts with shebangs are executable..........................Passed
check for broken symlinks............................(no files to check)Skipped
check toml...............................................................Passed
check xml............................................(no files to check)Skipped
check non-mkdocs yaml....................................................Passed
check mkdocs yaml........................................................Passed
fix end of files.........................................................Passed
detect destroyed symlinks................................................Passed
detect private key.......................................................Passed
detect aws credentials...................................................Passed
forbid submodules....................................(no files to check)Skipped
pretty format json.......................................................Passed
trim trailing whitespace.................................................Passed
ruff.....................................................................Passed
ruff-format..............................................................Passed
Detect secrets...........................................................Passed
check license header.....................................................Passed
```

novekm commented:
# PR #548: Security Scanning Enforcement Solution

## Problem Statement

I encountered significant challenges when implementing mandatory security scanning with `run_checkov()` before AWS resource operations:

1. **Inconsistent LLM Behavior**: LLMs would frequently skip the security scanning step entirely, creating resources without proper security validation.

2. **Client Compatibility Issues**: While my initial solution worked in Claude Desktop, it failed in Amazon Q CLI and Cursor with validation errors due to differences in how these clients handle function parameters and validation.

3. **Parameter Validation Challenges**: Making `security_check_result` a required parameter wasn't sufficient, as LLMs would sometimes provide invalid or fabricated results.

## My Initial Approach and Limitations

My initial implementation:
```python
async def create_resource(
    # ...other parameters...
    security_check_result: dict = Field(
        description='Result from run_checkov() to ensure security checks have been performed'
    ),
    # ...other parameters...
) -> dict:
    # Enforce that security_check_result comes from run_checkov
    if not skip_security_check:
        if not security_check_result or not isinstance(security_check_result, dict):
            raise ClientError(
                'You must call run_checkov() first and pass its result to this function'
            )
        # Verify security_check_result has required fields
        if 'passed' not in security_check_result:
            raise ClientError('Invalid security_check_result. You must call run_checkov() first')
```

This approach had several limitations:
- No validation that security checks were performed on the correct resource type
- No timestamp tracking to ensure checks were recent
- Inconsistent validation between create and update operations
- No clear workflow guidance for LLMs to follow

## My Comprehensive Security Scanning Solution

I've implemented a multi-layered approach that ensures security scanning is performed correctly across all operations and clients:

### 1. Infrastructure Code Generation Tool

I created the `generate_infrastructure_code()` tool as the cornerstone of my solution:

```python
async def generate_infrastructure_code(
    resource_type: str,
    properties: dict | None = None,
    identifier: str | None = None,
    patch_document: list | None = None,
    region: str | None = None,
    disable_default_tags: bool = False,
) -> dict:
    # Handles both creation and update scenarios
    # Returns CloudFormation template for security scanning
```

This tool is **critical** because:
- It provides a single entry point for both resource creation and update operations
- It generates consistent CloudFormation templates for security scanning
- It enforces a clear workflow that LLMs must follow
- It works across all clients (Claude Desktop, Amazon Q CLI, and Cursor)

A key insight from my testing was that **using Checkov directly in the same tool that needed its output didn't work**. LLMs would frequently skip the security scanning step or provide fabricated results. By decoupling the infrastructure code generation from the security scanning and resource creation steps, I created a proper dependency chain that enforces the steps the LLM can and cannot take. This separation is essential for ensuring the LLM follows the correct workflow.

### 2. Enhanced Security Validation

I developed a centralized `security_validator.py` that provides consistent validation:

```python
def validate_security_check_result(security_check_result, resource_type, skip_security_check=False):
    # Validates security check results
    # Ensures checks were performed on the correct resource type
    # Verifies checks are recent (timestamp validation)
    # Processes security check failures
```

### 3. Enhanced `run_checkov()` Function

I enhanced the `run_checkov()` function:

```python
async def run_checkov(
    content: Any,
    file_type: Literal['json', 'yaml', 'hcl'],
    framework: str | None = None,
    resource_type: str | None = None,  # Optional parameter
) -> dict:
    # Returns results with timestamp and resource type
    return {
        'passed': True/False,
        'failed_checks': [...],
        'passed_checks': [...],
        'summary': {...},
        'timestamp': timestamp,  # Added field
        'resource_type': resource_type,  # Added field
    }
```

### 4. Updated MCP Instructions

I updated the instructions to guide LLMs through the correct workflow for each operation type.

## Workflow Enforcement Details

I implemented security scanning enforcement through multiple complementary mechanisms that work together to ensure the correct workflow is followed:

### Create/Update Operations Workflow

When a user requests to create or update a resource, my solution enforces this sequence:

1. **Environment Check**: `check_environment_variables()` verifies AWS credentials and provides critical context to the LLM about the AWS environment. This is essential as it helps the LLM understand the AWS account context it's operating in.

2. **Session Info**: `get_aws_session_info()` retrieves account ID, region, etc. This provides the LLM with specific AWS account information that it can use when generating IAM policies and other resources. In testing, this significantly reduced insecure IAM policy generation (like using "AWS:":"*" for principals) because the LLM now has the actual account ID to use.

3. **Account Info**: `get_aws_account_info()` provides additional account details that further enhance the LLM's context. This contextual information is crucial for security as it guides the LLM to create properly scoped resources with appropriate permissions.

4. **Generate Infrastructure Code**: 
   ```python
   code = await generate_infrastructure_code(
       resource_type="AWS::S3::Bucket",
       properties={"BucketName": "my-bucket", ...}
   )
   ```

5. **Security Scan**: 
   ```python
   security_result = await run_checkov(
       content=code["cloudformation_template"],
       file_type="json",
       framework="cloudformation",
       resource_type="AWS::S3::Bucket"
   )
   ```

6. **Present Findings**: Security findings are shown to the user for approval.

7. **Create/Update Resource**: Only after approval, the resource is created/updated with the security check result.

The environment and account information functions are not just for validation - they provide essential context that guides the LLM to generate more secure resources from the start. This proactive approach significantly reduces security issues before they even reach the scanning stage.

### Security Analysis of Existing Resources

I've added security analysis capabilities to the read operations:

1. **Get Resource with Security Analysis**:
   ```python
   resource_info = await get_resource(
       resource_type="AWS::S3::Bucket",
       identifier="my-bucket",
       analyze_security=True
   )
   ```

2. **List Resources with Security Analysis**:
   ```python
   resources = await list_resources(
       resource_type="AWS::S3::Bucket",
       analyze_security=True,
       max_resources_to_analyze=5
   )
   ```

This allows users to ask questions like "Is my EC2 instance secure?" or "What's the security of my S3 buckets?" and get meaningful security analysis.

### Enforcement Mechanisms

I enforce the order through:

1. **Required Parameters**: `security_check_result` is required in resource operations.
2. **Security Validator**: `validate_security_check_result()` validates:
   - Security checks were performed
   - They were performed on the correct resource type
   - They were performed recently (timestamp validation)
   - They didn't find any high-severity issues
3. **Instructions**: MCP instructions guide the LLM through the workflow.
4. **AWS Context**: The environment and account information functions provide essential context that helps the LLM generate secure resources from the start.

## Amazon Q CLI Compatibility Issue

I discovered a specific issue with Amazon Q CLI that required additional attention:

### The Problem

The Amazon Q CLI was failing with a "Improperly formed request" validation error for every prompt, even simple ones like "hello". After investigation, I found that:

1. **Required Parameter Validation**: When `resource_type` was marked as `required=True` in the `run_checkov()` function, Q CLI's validation was more strict than Claude Desktop's. It was trying to validate this parameter even in simple conversations that don't involve security scanning.

2. **Context Size**: The validation error occurred because the context was becoming too large or improperly formatted. The required parameter validation was causing Q CLI to include validation information in every request.

### My Solution

I implemented a targeted fix that maintains security enforcement while resolving the Q CLI compatibility issue:

1. **Made `resource_type` Optional in `run_checkov()`**:
   ```python
   resource_type: str | None = Field(
       description='The AWS resource type being scanned (e.g., "AWS::S3::Bucket")',
       default=None,  # Changed from required=True
   )
   ```

2. **Moved Enforcement to Validation Logic**: Instead of relying on the function signature to enforce security scanning, I moved the enforcement to the validation logic in `security_validator.py`.

## Why My Solution Is Essential

My implementation is the **only solution** that meets all requirements:

1. **Client Compatibility**: Works across all MCP clients (Claude Desktop, Amazon Q CLI, and Cursor)
2. **Enforced Security**: Ensures security scanning is performed before any resource operation
3. **Correct Resource Validation**: Validates that security checks are performed on the correct resource type
4. **Timestamp Tracking**: Ensures security checks are recent
5. **Clear LLM Guidance**: Provides a clear workflow for LLMs to follow
6. **Security Analysis of Existing Resources**: Enables security scanning of existing resources
7. **Contextual Security**: The environment and account information functions provide critical AWS context to the LLM, which has been shown in testing to vastly reduce insecure IAM policy generation and other security issues

**No component can be removed** without breaking the solution:
- Without `generate_infrastructure_code()`, LLMs would revert to skipping security checks
- Without enhanced `run_checkov()`, resource type tracking would be lost
- Without centralized validation, consistency across operations would be compromised
- Without updated instructions, LLMs would not follow the correct workflow
- Without the environment and account information functions, the LLM would lack critical context for generating secure resources (especially IAM policies), leading to more security issues that would need to be caught by scanning

## Testing and Validation

I've tested my solution across all MCP clients:
- Claude Desktop: Works correctly with the new workflow
- Amazon Q CLI: Now properly enforces security scanning without validation errors
- Cursor: Successfully implements the security workflow

My testing specifically showed that providing AWS account context through the environment and account information functions significantly reduced the generation of insecure IAM policies. For example, when the LLM had access to the actual AWS account ID, it properly used that ID in IAM policy principals instead of using wildcards like "AWS:":"*", which is a common security issue.

This multi-layered approach ensures that security is addressed both proactively (through context and guidance) and reactively (through scanning and validation), creating a comprehensive security solution for AWS resource management.

novekm commented:
will fix tests, just wanted the new changes to be here in so folks can take a look. I was about to cut out about 2k lines of code. Everything in this most recent commit has to be in here to maintain the full security scanning functionality. Reasoning is outlined in my above comment, helped written by Amazon Q 😎

novekm commented:
Ok I think I fixed all of the tests that relate to this PR. Local pre-commit is passing all tests. I'm not sure why the pipeline is running checks for every single mcp server in the repo when this PR only touched the `cfn-mcp-server` directory, and I believe the readme. Is something wrong with the pipeline? Running tests on every mcp server seems like something that should only be triggered when a new release is cut for the entire project, not a PR for one mcp server
